maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -101,34 +83,91 @@ public boolean accept(File file) {
-            String downloadedPath = downloaded.getAbsolutePath();
+    private void extractFile(File downloaded, File toExtract, String downloadedPath) {
+        File tempFile = createTempFile(toExtract);
+        try {
+            if (downloadedPath.endsWith(".zip")) {
+                Spacelift.task(downloaded, UnzipTool.class).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.gz")) {
+                Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.bz2")) {
+                Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+            } else {
+                throw new IllegalArgumentException(
+                   "The distribution " + downloaded + " is compressed by unsupported format. "
+                      + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            }
+        } catch (ExecutionException ee) {
+            throw new IllegalStateException(
+               "Something bad happened when the file: " + downloadedPath + " was being extracted. "
+                  + "For more information see the stacktrace", ee);
+        }
+        if (!tempFile.delete()) {
+            log.warning("failed to delete temp directory: " + tempFile);
+        }
+    }
 
-            try {
-                if (downloadedPath.endsWith(".zip")) {
-                    Spacelift.task(downloaded, UnzipTool.class).toDir(toExtract).execute().await();
+    private File checkIfItIsAlreadyExtracted(final String downloadedZipMd5hash) {
 
-                } else if (downloadedPath.endsWith(".tar.gz")) {
-                    Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+        File[] dirs = new File(MAVEN_TARGET_DIR).listFiles(new FileFilter() {
 
-                } else if (downloadedPath.endsWith(".tar.bz2")) {
-                    Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+            @Override
+            public boolean accept(File file) {
+                return file.isDirectory() && downloadedZipMd5hash.equals(file.getName());
+            }
+        });
+        if (dirs != null && isExtractionFinished(dirs)) {
+            return dirs[0];
+        }
+        return null;
+    }
 
-                } else {
-                    throw new IllegalArgumentException(
-                            "The distribution " + downloaded + " is compressed by unsupported format. "
-                                    + "Supported formats are .zip, .tar.gz, .tar.bz2");
+    private boolean isExtractionFinished(File[] dirs) {
+        if (dirs.length > 0) {
+            File dir = dirs[0];
+            final File[] files = dir.listFiles(new FileFilter() {
+                @Override
+                public boolean accept(File file) {
+                    return file.isDirectory();
```

Comment:
if we have the marker file there then it is not necessary to check the directory I think...